src/xrpld/app/tx/detail/Batch.cpp

```diff
@@ -0,0 +1,316 @@
+            return temBAD_SIGNER;
+        }
+    }
+
+    if (ctx.tx.isFieldPresent(sfBatchSigners) &&
+        uniqueSigners.size() != ctx.tx.getFieldArray(sfBatchSigners).size())
+    {
+        JLOG(ctx.j.trace())
+            << "Batch: unique signers does not match batch signers.";
+        return temBAD_SIGNER;
+    }
+
+    return tesSUCCESS;
+}
+
+TER
+Batch::preclaim(PreclaimContext const& ctx)
+{
+    return tesSUCCESS;
+}
+
+TER
+Batch::doApply()
+{
+    Sandbox sb(&ctx_.view());
+    bool changed = false;
+    auto const flags = ctx_.tx.getFlags();
+
+    AccountID const outerAccount = ctx_.tx.getAccountID(sfAccount);
+
+    TER result = tesSUCCESS;
+    ApplyViewImpl& avi = dynamic_cast<ApplyViewImpl&>(ctx_.view());
+    OpenView subView(&sb);
+
+    auto const& txns = ctx_.tx.getFieldArray(sfRawTransactions);
+    bool const not3rdParty = std::any_of(
+        txns.begin(), txns.end(), [outerAccount](STObject const& txn) {
+            return txn.getAccountID(sfAccount) == outerAccount;
+        });
+
+    for (STObject txn : txns)
+    {
+        OpenView innerView(&subView);
+
+        STTx const stx = STTx{std::move(txn)};
+        auto const [ter, applied] =
+            ripple::apply(ctx_.app, innerView, stx, tapFAIL_HARD, ctx_.journal);
+
+        if (applied)
+            innerView.apply(subView);
+
+        changed = true;
+
+        // Add Inner Txn Metadata
+        STObject meta{sfBatchExecution};
+        std::string res = transToken(ter);
+        meta.setFieldVL(sfInnerResult, ripple::Slice{res.data(), res.size()});
+        meta.setFieldU16(sfTransactionType, stx.getTxnType());
+        if (ter == tesSUCCESS || isTecClaim(ter))
+            meta.setFieldH256(sfTransactionHash, stx.getTransactionID());
+        avi.addBatchExecutionMetaData(std::move(meta));
+
+        if (ter != tesSUCCESS)
+        {
+            // Atomic Revert on non tec failure
+            if (!isTecClaim(ter))
+            {
+                JLOG(ctx_.journal.trace()) << "Batch: Inner txn failed." << ter;
+                result = tecBATCH_FAILURE;
+                changed = false;
+                break;
+            }
+
+            if (flags & tfUntilFailure)
+            {
+                result = tesSUCCESS;
+                break;
+            }
+            if (flags & tfOnlyOne)
+            {
+                continue;
+            }
+            if (flags & tfAllOrNothing)
+            {
+                result = tecBATCH_FAILURE;
+                changed = false;
+                break;
+            }
+        }
+
+        if (ter == tesSUCCESS && flags & tfOnlyOne)
+        {
+            result = tesSUCCESS;
+            break;
+        }
+    }
+
+    // Apply SubView & PreviousFields
+    if (changed)
+    {
+        // Only required when not 3rd Party
+        if (not3rdParty)
+            ctx_.batchPrevious(avi);
+
+        ctx_.applyOpenView(subView);
+    }
+
+    sb.apply(ctx_.rawView());
```

Comment:
is sandbox always applied and never discarded? then calling it "sandbox" might be a bit misleading

Reply:
in fact, when i commented out line 312, all batch unit tests pass. So it seems like the sandbox isn't really applying anything since all the changes are in `subView`

Reply:
Did more testing, it seems like the the sandbox never has new state changes since `items_` appears to be always empty.

Reply:
Awesome thank you for reviewing I have removed it.